File: examples/quarter_truck/quarter_truck.cpp

#include "ecos/algorithm/fixed_step_algorithm.hpp"
#include "ecos/listeners/csv_writer.hpp"
#include "ecos/logger/logger.hpp"
#include "ecos/structure/simulation_structure.hpp"
#include "ecos/util/plotter.hpp"

using namespace ecos;

int main()
{
    set_logging_level(log::level::debug);

    const std::filesystem::path sspDir = std::string(DATA_FOLDER) + "/ssp/quarter_truck";
    const std::filesystem::path fmuDir = sspDir / "resources";

    try {
        simulation_structure ss;
        ss.add_model("chassis", fmuDir / "chassis.fmu");
        ss.add_model("ground", fmuDir / "ground.fmu");
        ss.add_model("wheel", fmuDir / "wheel.fmu");

        ss.make_connection<double>("chassis::p.e", "wheel::p1.e");
        ss.make_connection<double>("wheel::p1.f", "chassis::p.f");
        ss.make_connection<double>("wheel::p.e", "ground::p.e");
        ss.make_connection<double>("ground::p.f", "wheel::p.f");

        std::map<variable_identifier, scalar_value> map;
        map["chassis::C.mChassis"] = 400.0;
        ss.add_parameter_set("initialValues", map);

        const auto sim = ss.load(std::make_unique<fixed_step_algorithm>(1.0 / 100));

        auto csvWriter = std::make_unique<csv_writer>("results/quarter_truck_with_config.csv");
        const auto outputPath = csvWriter->output_path();
        csv_config& config = csvWriter->config();
        config.load(sspDir / "CsvConfig.xml");
        sim->add_listener("csv_writer", std::move(csvWriter));

        sim->init("initialValues");
        sim->step_until(5);

        sim->terminate();

        plot_csv(outputPath, sspDir / "ChartConfig.xml");
    } catch (const std::exception& ex) {

        log::err(ex.what());
    }
}

File: examples/quarter_truck/quarter_truck.py
from ecospy import *
from ecospy.plotter import *

from pathlib import Path

class MyListener(SimulationListener):

    def __init__(self, sim: EcosSimulation):
        self.sim = sim

    def pre(self, info: SimulationInfo):
        print(f"pre invoked @ t={info.time}")
        if info.time > 5:
            self.sim.remove_listener(self.name)
            print(f"Removed listener at t={info.time}")


def main():
    print(f"Ecoslib version: {EcosLib.version()}")

    EcosLib.set_log_level("debug")

    ssp_dir = (Path(__file__).parent.parent.parent / 'data' / 'ssp' / 'quarter_truck').resolve()
    result_file = f"results/python/quarter_truck.csv"

    with (EcosSimulation(ssp_path=f"{ssp_dir}", step_size=1.0 / 100)) as sim:

        sim.add_csv_writer(result_file, f"{ssp_dir}/LogConfig.xml")
        sim.add_listener("custom_listener", MyListener(sim))

        sim.init(parameter_set="initialValues")
        sim.step_until(time_point=10)
        sim.terminate()

    configs = [
        TimeSeriesConfig(
            title="Quarter-truck",
            y_label="Height[m]",
            identifiers=["chassis::zChassis"]),
        TimeSeriesConfig(
            title="Quarter-truck",
            y_label="Height[m]",
            identifiers=["wheel::zWheel"]),
        TimeSeriesConfig(
            title="Quarter-truck",
            y_label="Height[m]",
            identifiers=["ground::zGround"])
    ]
    plotter = Plotter(result_file, configs)
    plotter.show()


if __name__ == "__main__":
    main()
